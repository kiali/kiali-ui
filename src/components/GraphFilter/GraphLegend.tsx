import * as React from 'react';
import Draggable from 'react-draggable';
import { Card, Icon } from 'patternfly-react';
import { PfColors } from '../../components/Pf/PfColors';
import { style } from 'typestyle';
import { NestedCSSProperties } from 'typestyle/lib/types';

export interface GraphLegendProps {
  closeLegend: () => void;
}

export interface GraphLegendState {}

export interface ArrowProps {
  color: string;
  label: string;
  dashed: boolean;
}
const dashedTemplate = (color: string) => {
  return `repeating-linear-gradient(90deg, ${color},  ${color} 10px, white 10px, white 20px)`;
};

const Arrow = (props: ArrowProps) => {
  const line: NestedCSSProperties = { marginTop: '5px', width: '90px', height: '5px', float: 'left' };
  if (props.dashed) {
    line['backgroundImage'] = dashedTemplate(props.color);
  } else {
    line['backgroundColor'] = props.color;
  }
  const lineBase = style(line);
  const arrowStyle = style({ marginTop: 20, clear: 'both' });
  const arrowDiv = style({ width: '120px', float: 'left' });
  const labelStyle = style({ float: 'left', verticalAlign: 'middle', height: '20px' });
  return (
    <div className={arrowStyle}>
      <div className={arrowDiv}>
        <div className={lineBase} />
      </div>
      <div className={labelStyle}>{props.label}</div>
    </div>
  );
};

export default class GraphLegend extends React.Component<GraphLegendProps, GraphLegendState> {
  constructor(props: GraphLegendProps) {
    super(props);
  }

  render() {
    const dragHandlers = {};
    const cardStyle = style({ zIndex: 10, width: '20%', height: '200px' });
    return (
      <Draggable defaultPosition={{ x: 20, y: 600 }} {...dragHandlers}>
        <Card accented={true} className={cardStyle}>
          <Card.Heading>
            <Icon
              type="pf"
              name="close"
              aria-hidden="true"
              onClick={this.props.closeLegend}
              style={{ float: 'right' }}
            />
            <Card.Title>Traffic Legend</Card.Title>
          </Card.Heading>
          <Card.Body>
            <Arrow color={PfColors.Red100} label={'Over 10% Error'} dashed={false} />
            <Arrow color={PfColors.Orange400} label={'5 - 10% Error'} dashed={false} />
            <Arrow color={PfColors.Green400} label={'< 5% Error'} dashed={false} />
            <Arrow color={PfColors.Black} label={'Idle'} dashed={false} />
            <Arrow color={PfColors.Black} label={'No Traffic yet'} dashed={true} />
          </Card.Body>
        </Card>
      </Draggable>
    );
  }
}
