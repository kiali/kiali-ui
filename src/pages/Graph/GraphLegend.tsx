import * as React from 'react';
import { style } from 'typestyle';
import legendData, { GraphLegendItem, GraphLegendItemRow } from './GraphLegendData';
import { Button, Tooltip } from '@patternfly/react-core';
import CloseIcon from '@patternfly/react-icons/dist/js/icons/close-icon';
import { PFColors } from 'components/Pf/PfColors';
import { summaryFont, summaryTitle } from './SummaryPanelCommon';

export interface GraphLegendProps {
  closeLegend: () => void;
  className?: string;
  isMTLSEnabled: boolean;
}

const width = '190px';

export default class GraphLegend extends React.Component<GraphLegendProps> {
  render() {
    const legendBoxStyle = style({
      backgroundColor: PFColors.White,
      border: '1px #ddd solid',
      margin: '0 0 3.25em 0',
      overflow: 'hidden',
      overflowY: 'auto',
      padding: '1em 0.5em 1em 1em',
      zIndex: 3
    });

    const headerStyle = style({
      width: width
    });

    const bodyStyle = style({
      height: 'auto',
      width: width
    });

    const closeBoxStyle = style({
      float: 'right',
      margin: '-7px -5px 0 -10px'
    });

    return (
      <div className={legendBoxStyle} style={summaryFont}>
        <div className={`${headerStyle} ${summaryTitle}`}>
          <span>Legend</span>
          <span className={closeBoxStyle}>
            <Tooltip content="Close Legend">
              <Button id="legend_close" variant="plain" onClick={this.props.closeLegend}>
                <CloseIcon />
              </Button>
            </Tooltip>
          </span>
        </div>
        <div className={bodyStyle}>
          <div>{this.renderGraphLegendList(legendData)}</div>
        </div>
      </div>
    );
  }

  renderGraphLegendList(legendData: GraphLegendItem[]) {
    const legendColumnHeadingStyle = style({
      fontWeight: 'bold',
      paddingTop: '1.25em'
    });
    const aStyle = style({
      height: '100%'
    });

    return (
      <div className={aStyle}>
        {legendData.map((legendItem: GraphLegendItem) => (
          <div key={legendItem.title} className={legendColumnHeadingStyle}>
            {legendItem.title}
            {this.renderLegendRowItems(legendItem.data)}
          </div>
        ))}
      </div>
    );
  }

  renderLegendRowItems(legendData: GraphLegendItemRow[]) {
    return (
      <>{legendData.map((legendItemRow: GraphLegendItemRow) => GraphLegend.renderLegendIconAndLabel(legendItemRow))}</>
    );
  }

  static renderLegendIconAndLabel(legendItemRow: GraphLegendItemRow) {
    const keyWidth = '70px';

    const keyStyle = style({
      minWidth: keyWidth,
      width: keyWidth
    });

    const legendItemStyle = style({
      display: 'flex',
      flexDirection: 'row',
      padding: '5px 5px 0 5px'
    });

    const legendItemLabelStyle = style({
      fontWeight: 'normal'
    });

    return (
      <div key={legendItemRow.icon} className={legendItemStyle}>
        <span className={keyStyle}>
          <img alt={legendItemRow.label} src={legendItemRow.icon} />
        </span>
        <span className={legendItemLabelStyle}>{legendItemRow.label}</span>
      </div>
    );
  }
}
