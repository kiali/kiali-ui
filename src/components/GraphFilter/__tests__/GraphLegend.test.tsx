import * as React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import GraphLegend from '../GraphLegend';

describe('GraphLegend test', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<GraphLegend closeLegend={jest.fn()} isMTLSEnabled={false} />);
    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
