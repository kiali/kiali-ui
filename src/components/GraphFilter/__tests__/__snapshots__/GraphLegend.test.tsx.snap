// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphLegend test should render correctly 1`] = `""`;
